iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```diff
@@ -79,105 +77,214 @@ static llvm::cl::opt<int> defaultWorkgroupTileSize(
-        if (auto nativeVectorSizeAttr =
-                config.getAs<IntegerAttr>("native_vector_size")) {
-          if (int64_t nativeVectorSizeVal = nativeVectorSizeAttr.getInt()) {
-            nativeVectorSizeInBytes = nativeVectorSizeVal;
-          }
-        }
-      }
+/// Looks for the `native_vector_size` attribute in the hal.executable.variant
+/// op.
+static Optional<int64_t> getNativeVectorSizeInBytes(FuncOp entryPointFn) {
+  auto variantOp =
+      entryPointFn->getParentOfType<IREE::HAL::ExecutableVariantOp>();
+  if (!variantOp) return llvm::None;
+  IREE::HAL::ExecutableTargetAttr targetAttr = variantOp.target();
+  if (!targetAttr) return llvm::None;
+  auto config = targetAttr.getConfiguration();
+  if (!config) return llvm::None;
+  auto nativeVectorSizeAttr = config.getAs<IntegerAttr>("native_vector_size");
+  if (!nativeVectorSizeAttr) return llvm::None;
+  int64_t nativeVectorSizeVal = nativeVectorSizeAttr.getInt();
+  if (!nativeVectorSizeVal) return llvm::None;
+  return nativeVectorSizeVal;
+}
+
+/// Return the type length in bytes. Looks throuigh all the interface binding
+/// ops to see the ABI types and guess-timates the type size to use.
+static unsigned getReferenceTypeLengthInBytes(FuncOp entryPointFn) {
+  unsigned referenceTypeLengthInBytes = 4;
+  entryPointFn.walk([&](IREE::HAL::InterfaceBindingSubspanOp subSpanOp) {
+    Type type = subSpanOp.getResult().getType();
+    Type elementType = TypeSwitch<Type, Type>(type)
+                           .Case<ShapedType, IREE::Flow::DispatchTensorType>(
+                               [&](auto shapedType) -> Type {
+                                 if (shapedType.getRank() > 0) {
+                                   return shapedType.getElementType();
+                                 }
+                                 return nullptr;
+                               })
+                           .Default([&](Type t) -> Type { return nullptr; });
+    if (!elementType || !elementType.isIntOrFloat()) return;
+    unsigned typeWidthInBytes =
+        std::max<unsigned>(elementType.getIntOrFloatBitWidth() / 8, 1);
+    referenceTypeLengthInBytes =
+        std::min<unsigned>(referenceTypeLengthInBytes, typeWidthInBytes);
+  });
+  return referenceTypeLengthInBytes;
+}
+
+static SmallVector<int64_t> getDefaultWorkloadPerWorkgroup(
+    ArrayRef<TiledLoopInfo> tiledLoops,
+    ArrayRef<int64_t> nativeVectorSizeInElements) {
+  if (tiledLoops.empty()) {
+    return {};
+  }
+  assert(tiledLoops.size() == nativeVectorSizeInElements.size());
+  int64_t useDefaultWorkgroupTileSize = defaultWorkgroupTileSize;
+  unsigned maxDim = 0;
+  for (auto tiledLoop : tiledLoops) {
+    maxDim = std::max<unsigned>(tiledLoop.distributionDim, maxDim);
+  }
+  SmallVector<int64_t> workloadPerWorkgroup(maxDim + 1, 1);
+  if (tiledLoops.size() == 1) {
+    useDefaultWorkgroupTileSize *= 2;
+  } else if (tiledLoops.size() == 3) {
+    useDefaultWorkgroupTileSize /= 2;
+  }
```

Comment:
This is not trivial to understand why we choose it this way, maybe add a comment on it?

Reply:
This was kind of hacky. Dropped it.